pkg/packageinstall/packageinstall_downgrade_test.go

```diff
@@ -0,0 +1,257 @@
+		_, err := ip.Reconcile()
+		assert.Nil(t, err)
+
+		assert.Equal(t, pkgingv1alpha1.PackageInstallStatus{
+			Version:              "1.0.0",
+			LastAttemptedVersion: "2.0.0",
+			GenericStatus: v1alpha1.GenericStatus{
+				ObservedGeneration: 0,
+				Conditions: []v1alpha1.AppCondition{{
+					Type:    "ReconcileFailed",
+					Status:  "True",
+					Reason:  "",
+					Message: "Error (see .status.usefulErrorMessage for details)",
+				}},
+				FriendlyDescription: "Reconcile failed: Error (see .status.usefulErrorMessage for details)",
+				UsefulErrorMessage:  "Stopped installing matched version '1.0.0' since last attempted version '2.0.0' is higher",
+			},
+		}, getPackageInstall(t, appClient, "instl-pkg").Status)
+
+		assert.Equal(t, "ver-existing", getApp(t, appClient, "instl-pkg").Spec.Fetch[0].ImgpkgBundle.Image)
+	})
+
+	t.Run("succeeds when trying to install lower version but is allowed to downgrade", func(t *testing.T) {
+		pkgInstall := &pkgingv1alpha1.PackageInstall{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "instl-pkg",
+				Annotations: map[string]string{
+					"packaging.carvel.dev/downgradable": "",
+				},
+			},
+			Spec: pkgingv1alpha1.PackageInstallSpec{
+				PackageRef: &pkgingv1alpha1.PackageRef{
+					RefName: "expected-pkg",
+					VersionSelection: &versions.VersionSelectionSemver{
+						Constraints: "1.0.0",
+					},
+				},
+			},
+			Status: pkgingv1alpha1.PackageInstallStatus{
+				LastAttemptedVersion: "2.0.0", // higher than 1.0.0
+			},
+		}
+
+		pkgClient := fakeapiserver.NewSimpleClientset(&pkg1, &pkg2)
+		appClient := fakekappctrl.NewSimpleClientset(pkgInstall, existingApp)
+		coreClient := fake.NewSimpleClientset()
+
+		ip := NewPackageInstallCR(pkgInstall, log, appClient, pkgClient, coreClient)
+		_, err := ip.Reconcile()
+		assert.Nil(t, err)
+
+		assert.Equal(t, pkgingv1alpha1.PackageInstallStatus{
+			Version:              "1.0.0",
+			LastAttemptedVersion: "1.0.0",
+			GenericStatus: v1alpha1.GenericStatus{
+				ObservedGeneration: 0,
+				Conditions: []v1alpha1.AppCondition{{
+					Type:   "Reconciling",
+					Status: "True",
+				}},
+				FriendlyDescription: "Reconciling",
+			},
+		}, getPackageInstall(t, appClient, "instl-pkg").Status)
+
+		assert.Equal(t, "ver-1.0.0", getApp(t, appClient, "instl-pkg").Spec.Fetch[0].ImgpkgBundle.Image)
+	})
+}
+
+func getPackageInstall(t *testing.T, clientset *fakekappctrl.Clientset, name string) *pkgingv1alpha1.PackageInstall {
+	gvr := schema.GroupVersionResource{"packaging.carvel.dev", "v1alpha1", "packageinstalls"}
+	obj, err := clientset.Tracker().Get(gvr, "", name)
+	assert.Nil(t, err)
+	assert.NotNil(t, obj)
+	return obj.(*pkgingv1alpha1.PackageInstall)
+}
+
+func getApp(t *testing.T, clientset *fakekappctrl.Clientset, name string) *v1alpha1.App {
+	gvr := schema.GroupVersionResource{"kappctrl.k14s.io", "v1alpha1", "apps"}
+	obj, err := clientset.Tracker().Get(gvr, "", name)
+	assert.Nil(t, err)
+	assert.NotNil(t, obj)
```

Comment:
I think `require.NotNil` would be helpful here to avoid nil being returned

Reply:
fixed.